### PR TITLE
Support Env registry and create from string

### PR DESCRIPTION
**Pull request by sduzh**
## What type of PR is this：
- [ ] bug
- [ ] feature
- [x] enhancement
- [ ] others

## Problem Summary(Required) ：

Support Env registry and create an instance of Env from string

### CODE REVIEW

File: be/src/env/env.cpp [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+#include <regex>
+#include <type_traits>
+
+#include "env/env_hdfs.h"
+#include "env/env_posix.h"
+#include "env/env_s3.h"
+
+namespace starrocks {
+
+class EnvRegistry {
```
Comment (chaoyli) [inline be/src/env/env.cpp:16]: Is there necessary to use a new Class?

File: be/src/env/env.cpp [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+public:
+    using FactoryFunc = Env::FactoryFunc;
+
+    static EnvRegistry& Instance() {
```
Comment (chaoyli) [inline be/src/env/env.cpp:20]: Forbidden the copy-constructor and assignment operator.

File: be/src/env/env.h [hunk 3]
```diff
@@ -41,6 +44,12 @@ class Env {
     Env() = default;
     virtual ~Env() = default;
 
+    static void Register(std::string_view pattern, FactoryFunc func);
+
+    static StatusOr<std::unique_ptr<Env>> CreateUniqueFromString(std::string_view uri);
```
Comment (chaoyli) [inline be/src/env/env.h:49]: CreateFromString is a better name.

File: be/src/env/env.h [hunk 4]
```diff
@@ -41,6 +44,12 @@ class Env {
...
+
+    static StatusOr<std::unique_ptr<Env>> CreateUniqueFromStringOrDefault(std::string_view uri);
```
Comment (chaoyli) [inline be/src/env/env.h:51]: CreateFromStringOrDefault is a better name.

File: be/src/env/env.h (hunk 3 above)
Comment (sduzh) [reply be/src/env/env.h:49]: It's highly possible to return a std::shared_ptr<Env> in the future.

File: be/src/env/env.cpp (hunk 1 above)
Comment (sduzh) [reply be/src/env/env.cpp:16]: Any suggestion?

File: be/src/env/env.cpp (hunk 2 above)
Comment (sduzh) [reply be/src/env/env.cpp:20]: updated

File: be/src/env/env.cpp (hunk 1 above)
Comment (sduzh) [reply be/src/env/env.cpp:16]: I don't want to add some data structures in the abstract class Env.

**Comment by wanpengfei-git** [comment]
#### [BE PR Coverage check]


:heart_eyes: **pass** : 0 / 0 (0%)

File: be/src/env/env.cpp [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+        static EnvRegistry instance;
+        return instance;
+    }
+
+    // Disallow copy ctor and copy assignment
```
[the remark this answers is not in the document]
Comment (sduzh) [reply be/src/env/env.cpp]: updated

**Review by dirtysalt — DISMISSED** [review]

**Review by wyb — DISMISSED** [review]

**Comment by sduzh** [comment]
run starrocks_be_unittest

**Review by dirtysalt — APPROVED** [review]

**Review by wyb — APPROVED** [review]

**Pull request merged**